File: MTG Deck Editor/src/editor/database/characteristics/Rarity.java
package editor.database.characteristics;

/**
 * This enum represents a rarity a Magic: The Gathering card can have.  Rarities
 * are basically Strings, so they implement {@link CharSequence}.  All of the
 * implemented methods operate on a Rarity's String representation from
 * {@link #toString()}.
 * 
 * @author Alec Roelke
 */
public enum Rarity implements CharSequence
{
	/**
	 * Rarity for basic lands.
	 */
	BASIC_LAND("Basic Land"),
	/**
	 * Common rarity.
	 */
	COMMON("Common"),
	/**
	 * Uncommon rarity.
	 */
	UNCOMMON("Uncommon"),
	/**
	 * Rare rarity.
	 */
	RARE("Rare"),
	/**
	 * Mythic rare rarity.
	 */
	MYTHIC_RARE("Mythic Rare"),
	/**
	 * "Special" rarity, such as timeshifted.
	 */
	SPECIAL("Special");
	
	/**
	 * Create a rarity from a shorthand character.
	 * 
	 * @param rarity Character to create a Rarity from
	 * @return a Rarity representing the specified shorthand character
	 * @throws IllegalArgumentException if a Rarity cannot be created from the specified character
	 */
	public static Rarity parseRarity(char rarity)
	{
		rarity = Character.toLowerCase(rarity);
		switch (rarity)
		{
		case 'c':
			return COMMON;
		case 'u':
			return UNCOMMON;
		case 'r':
			return RARE;
		case 'm':
			return MYTHIC_RARE;
		case 's':
			return SPECIAL;
		case 'b':
			return BASIC_LAND;
		default:
			throw new IllegalArgumentException("Illegal rarity shorthand");
		}
	}
	
	/**
	 * Create a Rarity from the specified String.
	 * 
	 * @param rarity String to create a Rarity from
	 * @return a Rarity representing the specified String
	 * @throws IllegalArgumentException if a Rarity cannot be created from the String
	 */
	public static Rarity parseRarity(String rarity)
	{
		for (Rarity r: Rarity.values())
			if (rarity.equalsIgnoreCase(r.rarity) || rarity.equalsIgnoreCase(String.valueOf(r.shorthand())))
				return r;
		if (rarity.contains("mythic"))
			return MYTHIC_RARE;
		else if (rarity.contains("basic"))
			return BASIC_LAND;
		else
			throw new IllegalArgumentException("Illegal rarity string \"" + rarity + "\"");
	}
	
	/**
	 * String representation of this Rarity.
	 */
	private final String rarity;
	
	/**
	 * Create a new Rarity.
	 * 
	 * @param rarity String representation of the new Rarity.
	 */
	private Rarity(final String rarity)
	{
		this.rarity = rarity;
	}
	
	@Override
	public char charAt(int index)
	{
		return rarity.charAt(index);
	}

	@Override
	public int length()
	{
		return rarity.length();
	}

	/**
	 * Get the shorthand character for this Rarity.
	 * 
	 * @return A shorthand character representing this Rarity.
	 */
	public char shorthand()
	{
		switch(this)
		{
		case COMMON:
			return 'C';
		case UNCOMMON:
			return 'U';
		case RARE:
			return 'R';
		case MYTHIC_RARE:
			return 'M';
		case SPECIAL:
			return 'S';
		case BASIC_LAND:
			return 'B';
		default:
			return '\0';
		}
	}

	@Override
	public CharSequence subSequence(int start, int end)
	{
		return rarity.subSequence(start, end);
	}

	@Override
	public String toString()
	{
		return rarity;
	}
}

File: src/main/scala/editor/database/attributes/Rarity.scala
package editor.database.attributes

/**
 * Functions for parsing [[Rarity]] from strings and characters.
 * @author Alec Roelke
 */
object Rarity {
  /** @return the [[Rarity]] whose shorthand matches the given character, or None if there isn't one. */
  def parse(rarity: Char) = values.find(_.shorthand == rarity.toUpper)

  /** @return the [[Rarity]] that matches the given string, or None if there isn't one. */
  def parse(rarity: String) = {
    val lower = rarity.toLowerCase
    val names = values.map((r) => r -> r.unique.toLowerCase)
    names.reverse.collect{ case (r, n) if lower == r.shorthand.toString || lower.contains(n) => r }.headOption
  }
}

/**
 * A possible rarity that a card can have, which approximately represents its frequency in appearing in booster packs.
 * 
 * @constructor create a new rarity
 * @param toString displayable string representation
 * @param unique internal representation to use for parsing
 * @param shorthand single-character representation
 * 
 * @author Alec Roelke
 */
enum Rarity(override val toString: String, val unique: String, val shorthand: Char) extends Ordered[Rarity] {
  override def compare(that: Rarity) = ordinal - that.ordinal

  /** Formerly used for basic lands, which are now [[COMMON]]. */
  @deprecated case BASIC_LAND extends Rarity("Basic Land", "basic", 'B')
  /** Common rarity. Appears 10-11 times per normal booster pack, including the basic land. */
  case COMMON extends Rarity("Common", "common", 'C')
  /** Uncommon rarity. Appears 3-4 times per normal booster pack. */
  case UNCOMMON extends Rarity("Uncommon", "uncommon", 'U')
  /** Rare rarity. Appears once per normal booster pack. */
  case RARE extends Rarity("Rare", "rare", 'R')
  /** Mythic rare rarity. Has a 1/8 chnce of replacing the rare in a normal booster pack. */
  case MYTHIC_RARE extends Rarity("Mythic Rare", "mythic", 'M')
  /** Special rarity. Some expansions have additional cards that might replace a common in a booster pack. */
  case SPECIAL extends Rarity("Special", "special", 'S')
  /** Bonus rarity. Used for extremely rare cards in certain online booster packs. */
  case BONUS extends Rarity("Bonus", "bonus", 'O')
  /** Rarity couldn't be determined. */
  case UNKNOWN extends Rarity("Unknown", "unknown", 0)
}